src/Security/Security.php

```diff
@@ -498,9 +498,35 @@ public function ping()
      */
     public function logout($redirect = true)
     {
-        $this->extend('beforeMemberLoggedOut');
+        // Check pre-logout process
+        if ($response = $this->preLogout()) {
+            return $response;
+        }
+
         $member = static::getCurrentUser();
 
+        // If the user doesn't have a security token, show them a form where they can get one.
+        // This protects against nuisance CSRF attacks to log out users.
+        if ($member && !SecurityToken::inst()->checkRequest($this->getRequest())) {
```

Comment:
This should be in the LogoutHandler, for consistency and handled below, where the authenticators are checked and validated for the required supportedService.